### PR TITLE
Wait until response body or error body received to process request

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
This fixes issue reported in #4684. Basically, in situations where the identity used to call `s3.getObject()` has permissions to successfully perform a `HeadObject` but not a `GetObject` (i.e. this results in an error response), the S3 client thinks the the call was actually successful (because `HeadObject` was successful) and returns the error response content as the object content.

## Modifications
- Don't rely on `onResponseHeaders` because these are not indicative if whether the request succeeded. Move calls to `responseHandler.onHeanders()` to `onResponseBody` and `onFinished`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java [hunk 1]
```diff
@@ -63,14 +65,13 @@ public void onResponseHeaders(int statusCode, HttpHeader[] headers) {
...
         }
-
         respBuilder.statusCode(statusCode);
-        responseHandler.onHeaders(respBuilder.build());
-        responseHandler.onStream(responsePublisher);
     }
 
     @Override
     public int onResponseBody(ByteBuffer bodyBytesIn, long objectRangeStart, long objectRangeEnd) {
+        initiateResponseHandling(respBuilder.build());
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:73]: If there's no response payload, onResponseBody will not be invoked, so should we call initiateResponseHandling in `onFinished` for successful response as well?

Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:73]: Ah right, makes sense.

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java [hunk 2]
```diff
@@ -142,6 +148,16 @@ private void handleError(S3FinishedResponseContext context) {
         }
     }
 
+    private void initiateResponseHandling(SdkHttpResponse response) {
+        if (!responseHandlingInitiated) {
+            responseHandlingInitiated = true;
+            responseHandler.onHeaders(response);
+            responseHandler.onStream(responsePublisher);
+        } else {
+            log.warn(() -> "Attempted to initiate response handling more than once. Ignoring this attempt.");
```
[the remark this answers is not in the document]
Comment (dagnir) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:157]: +1

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java [hunk 3]
```diff
@@ -142,6 +148,14 @@ private void handleError(S3FinishedResponseContext context) {
         }
     }
 
+    private void initiateResponseHandling(SdkHttpResponse response) {
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:151]: Can we add a comment to mention why we are doing it like this?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java [hunk 4]
```diff
@@ -60,17 +62,13 @@ public S3CrtResponseHandlerAdapter(CompletableFuture<Void> executeFuture,
...
-
-        respBuilder.statusCode(statusCode);
-        responseHandler.onHeaders(respBuilder.build());
-        responseHandler.onStream(responsePublisher);
+        populateSdkHttpResponse(initialHeadersResponse, statusCode, headers);
     }
 
     @Override
     public int onResponseBody(ByteBuffer bodyBytesIn, long objectRangeStart, long objectRangeEnd) {
+        initiateResponseHandling(initialHeadersResponse.build());
```
Comment (zoewangg) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:70]: Can we add a comment to mention why this is invoked here?

**Review by zoewangg — APPROVED** [review]

**Comment by zoewangg** [comment]
Can we also add motivation for this change and change log entry?

**Comment by dagnir** [comment]
> Can we also add motivation for this change and change log entry?

Done

**Pull request merged**